Project Memberships

brec9824 opened the pull request:
Added a member page to projects which requires access from admin of the project.
Features:
- Request access to project
- Admin add user
- Admin deny user
- Editable banner, mission statement, custom tab on project member page
- Kicking users from project

Also fixes #135, #127, #114, #99, #12, #11, #10 (Deleting account I setup some placeholder profile routes and controller)

Ready for review @uidaho/lambda 

Notes: added chat to the bottom right hand side of the project members page

Commit:
Fixed forbidden page

Overloaded the forbidden handler to bring up the 403 page. Also edited it so if the user is not logged in and clicks submit comment the loggin window mini window pops up.

Commit:
Add member table and relationships

Added a member table for getting members of a project.

Commit:
Added basic project membership

Added db tables for project members and project requests. Added elequant relationships and routes that were required. Lastly added requirements to view files.

Commit:
Removed buttons that project owner shouldnt see

Made it so the project owner cant see the follow and join buttons.
Also made it so when a user creates a project they are also made a member of the project.

Commit:
Added basic accepting and denying of project requests

Current users requesting to be a member of a project are displayed at the file list page. User's requests can be accepted and denied by someone already a member of the project. Added new routes to handle these requests and tweaked the editor view.

Commit:
Added new members only page

Now users of a project can go to a page that list admins, users, files, current progress, and is private from non members.

Commit:
Admin powers added

Added the ability for admins to promote others to admin and only the project owner can demote users from admin. Fixed some css issues and added some more funtionality to members page.

Commit:
Updated how admins are retrieved

Admins are know grabed off of a project object.

Commit:
Fixed Relationship between user and project

Now you can use $project->user to get the user that created the current project.

Commit:
Updated relationships and queries

Changed so queries are consistant.

Commit:
Removed code that was causing page error

Forgot to remove part of a replaced piece of code.

Commit:
Added members page creation to create a project

Create a project now includes the text box and inputs needed for the project members page.

Commit:
Removed members page creation from project create page

Added custom requests for the editable parts of the members page. Added routes for editing member page parts. Added to project table to store the editable members page data.

Commit:
Fixed edits not working

Editing now works and added some more css formating.

Commit:
Added removing member and edited file tab

Made it so admins can kick non admin users from the project. The file tab now includes create file and the search bar. Tweaked some of the css so the data inside of tabs are staying within their tabs. Also moved some of the routes to a project members controller. So all things located on the public project page are located in the project controller and everything located on the project private members page is located on the new controller.

Commit:
Make creator of a project auto follow it

Fixes #135 by making the project creator auto follow thier own projects.

Commit:
Added ability to delete a user

Added profile controller as a place holder until the real profile section gets added. Added profile routes, added missing User relationships, and added delete method on user to delete all items connected to a user. Lastly also added a policy for profile.

Commit:
Added project member policy for autherization checks

Now users are checked for permissions server side before handling admin changes, join requests, and kicking users.

Commit:
Finished member page authentication

Added the last few policies to do validation checking for members page. Also added a premade banner to when the project is created and it is chosen from a random set of premade banners.

Commit:
Added final validation checks

Added final validation checks on the project controller and added appropriate comments for methods.

Commit:
Deleting a project deletes all connected data.

Deleting a project now deletes connected comments, requests, members, followers, files, and images.

Commit:
Updated tests

Tests are now updated to work with the changes to the website.

Commit:
Fixed issue with test

Updated some of the tests landing pages.

Commit:
Fixed routes

Route was redirecting to wrong page.

Commit:
Fixed rebase error

Commit:
Deleted old members view

View has been replaced elsewhere.

Commit:
Added chat into members page

Chat is now located at the bottom right of the members page.

Commit:
Moved that chat around

Chat has been moved on the the edit file page and is in the bottom right. Moved the chat over a little on the project page and made in open by default.

Commit:
Changed chat input size

Chat input size now matches the size of the rest of the chat box text.

Commit:
Added back in character counts

Added character counts back into project create and added it to forms on members page.

Commit:
Updated project table

Updated project table with a seperate table for adding to it. Adjusted the chat so it didnt add extra width to the page. Also updated a the editor list page.

Commit:
Fixed text spilling and chat

Fixed comments and project body text spilling out of their containers. Also fixed issue of chat box scaling really small.

brec9824 commented:
Ok @backburn @rickbau5 fixed the database migration so it adds to old tables rather than editing them. Ready for review.

Commit:
Reduced image sizes

Reduced the sizes of the banner images so they werent so big.

backburn commented:
https://squireproject.com/project/Project%20One
Join Project and Follow buttons show up when not logged in, but do not show up when logged in.
Really, all buttons should be available or context aware. If you don't have permission it should ask you to login etc...

brec9824 commented:
Join project and follow buttons show when not logged in or not the owner of the project. If you are the owner of the project the those two buttons are hidden since the owner doesnt need them.

backburn commented:
im not the owner of the projkect, project->user_id for that project is 0.

Im thinking you are relying on a record being present in the db in all the project_ tables for each user?

backburn commented:
In this case, if there is no row in the table for the user, it should create one.

Commit:
Added button to get back to project member page

Added a button on the file editor and file list pages that links to the project member page.

brec9824 commented:
@backburn Im a little confused how to replicate what you did. Can you tell me what you did to have that happen. Because I tried nuking my database and seeding it with some projects. Then registering a account and checking one of the projects and the buttons were there.

backburn commented:
my account was already there, php artisan migrate worked fine. So, there are no records for my user id in the project_members or project_followers. I believe you are depending on there being an record in there, it should try or create instead, so this does not happen.

brec9824 commented:
@backburn Ok gotcha. I see a problem this will cause without wiping everyone's database. All previous projects will become inaccessible to anyone because all projects become connected to a user through memberships. I would have to detect all users with projects created and create memberships for them also.

rickbau5 commented on app/Http/Controllers/ProjectMembersController.php:
```diff
@@ -0,0 +1,175 @@
+        $this->authorize('authToAddDeny', [$project, $user->id]);
+
+        $project->deleteMembershipRequest($user->id);
+        //Session::flash('membership_request_success', 'You are no longer requesting membership to this project.');
```
comments in this file?

rickbau5 commented:
Looks good to merge to me. I'm fine with this nuking the databases if need be. It is a dev environment after!

brec9824 commented:
Ok will merge after a fix one thing I just noticed.

Commit:
Fixed javascript that wasnt working

Java script to hide edit buttons wasnt working so moved it over to php to hide the buttons for non admins.

Commit:
Small change

brec9824 commented:
Just so everyone knows you will have nuke your database then remigrate and also might require doing a composer update.
